[RSDK-8405] [RSDK-8408] Update README to include instructions for FFC-3P and Yolo Detection Network (#54)
viamrobotics/viam-camera-oak: README.md

Other OAK family cameras may work with the either component model, but are not officially supported and tested.

### Services
* `viam:luxonis:yolo-detection-network` - Configures [a DepthAI Yolo Detection Network](https://docs.luxonis.com/software/depthai-components/nodes/yolo_detection_network/) on the OAK device pipeline.

## How to use on your own machine
> [!NOTE]  
> Before configuring your resource, you must create a new smart machine. See [here](https://docs.viam.com/use-cases/configure/#configure-a-machine) for how to do that and more on the setup and configuration process as a whole.

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com/).

[Add `camera` / `oak`,  `camera` / `oak-d` or `camera` / `oak-ffc-3p` to your machine](https://docs.viam.com/configure/#components).
Enter a name for your camera and click **Create**.

To configure the vision service, click on the **Services** subtab and click **Create service**.

On the new resource config panel, modify the attributes JSON in the **Attributes** box. If you are confused which attributes to supply, consult the per-model attribute configuration guides below.

> ## Attribute Configuration Guides

Reviewer: ```suggestion
## Attribute configuration guides
```

### Configuring the `oak-d` component model

> #### Example Configuration

Reviewer: ```suggestion
#### Example configuration
```

Below is an example JSON of an `oak-d` component's attributes:
```json
{
  "name": "my-oak",
  "attributes": {
    "sensors": ["color", "depth"],
    "width_px": 640,
    "height_px": 480,
    "frame_rate": 30,
    "device_info": "<mxid-or-ip-address-or-usb-port-name>"
  },
  "namespace": "rdk",
  "type": "camera",
  "model": "viam:luxonis:oak-d"
}
```

#### Attributes

> The following attributes are available for the `oak-d` component:

Reviewer: ```suggestion
### `oak-d` component model

The following attributes are available for the `oak-d` camera component:
```

| `width_px` | int | Optional | Width in pixels of the images output by this camera. Default: `1280` |
| `height_px` | int | Optional | Height in pixels of the images output by this camera. Default: `720` |
| `frame_rate` | int | Optional | The frame rate the camera will capture images at. Default: `30` |
| `device_info` | string | Optional | Physical device identifier to connect to a specific OAK camera connected to your machine. If not specified, the module will pick the first device it detects. `device_info` can be a MXID, usb port path, or IP address. [See DepthAI documentation for more details](https://docs.luxonis.com/software/depthai/examples/device_information#Device%20information). |

> [!NOTE]  
> Higher resolutions may cause out of memory errors. See Luxonis documentation [here](https://docs.luxonis.com/projects/api/en/latest/tutorials/ram_usage/.).

> ### Configuring the `oak-ffc-3p` component model

Reviewer: ```suggestion
### `oak-ffc-3p` component model
```
To match

> #### Example Configuration

Reviewer: ```suggestion
#### Example configuration
```

| `type` | str | Required | The type of the sensor: "color" or "depth", corresponding to whether the sensor on the respective socket is a color or mono depth camera. |
| `width_px` | int | Optional | Width in pixels of the images output by this camera. Default: `1280` |
| `height_px` | int | Optional | Height in pixels of the images output by this camera. Default: `720` |
| `frame_rate` | int | Optional | The frame rate the camera will capture images at. Default: `30` |
| `color_order` | str | Optional | The color order of the output frames (used for color sensor type only): "rgb" or "bgr". Default: `rgb` |
| `interleaved` | bool | Optional | Whether or not output frames should be stored in an interleaved format. Default: `false` |

> [!NOTE]  
> Higher resolutions may cause out of memory errors. See Luxonis documentation [here](https://docs.luxonis.com/projects/api/en/latest/tutorials/ram_usage/.).

> ### Configuring the `yolo-detection-network` service model

Reviewer: ```suggestion
### `yolo-detection-network` service model
```
to match

> #### Example Configuration

Reviewer: ```suggestion
#### Example configuration
```

```json
{
  "cam_name": "my-oak",
  "input_source": "color",
  "num_threads": 2,
  "num_nce_per_thread": 1,
  "yolo_config": {
    "blob_path": "<local-path-to-yolo-model>",
    "label_map": ["person", "apple", "orange", "tv", "etc."],
    "confidence_threshold": 0.5,
    "iou_threshold": 0.5,
    "anchors": [0, 0, 1, 1, 2, 2, 3, 3, 4, 4],
    "anchor_masks": {
      "side26": [1, 2, 3],
      "side13": [3, 4, 5]
    },
    "coordinate_size": 4
  }
}
```

#### Attributes

The following attributes are available for the `yolo-detection-network` service:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `cam_name` | string | Required | The name of the Viam OAK camera component to set up the service on. |
| `input_source` | string | Required | The socket name i.e. "cam_a", "cam_b", "cam_c" to get neural network input frames from. Can also be set as "color" to select the primary color sensor on the underlying OAK component. |
| `num_threads` | int | Optional | How many threads should the DepthAI node use to run the network. |
| `num_nce_per_thread` | int | Optional | How many Neural Compute Engines should a single thread use for inference. |

> | `yolo_config` | struct | **Required** | A struct mapping of strings to values representing the sub-configuration for the YOLO model. |

Reviewer 2: Are there defaults here?

Author: Missed these good catch

Author: Done

| `labels` | list[string] | Required | A list of strings representing each label the YOLO model can output. Order matters here as YOLO models use label index mapping to retrieve string labels. |
| `confidence_threshold` | float | Required | The minimum confidence level required for a detection to be considered valid. Default `0.5`. |
| `iou_threshold` | float | Required | The Intersection Over Union (IOU) threshold used for non-max suppression to filter out overlapping bounding boxes. Default `0.5`. |
| `anchors` | list[float] | Optional | A list of floats representing the anchor boxes used by the YOLO model. The values should be in pairs representing width and height. Default: `[]` |
| `anchor_masks` | dict | Optional | A dictionary where keys are strings (e.g., "side26", "side13") representing different scales, and values are lists of integers representing the anchor indices used at each scale. Default: `{}` |
| `coordinate_size` | int | Optional | The number of coordinates used for each bounding box. Typically, this is 4 for (x, y, width, height). Default `4`. |

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```

Reviewer: Might want to add some actual next steps here besides debugging, like describing how to test the camera in the viam app? quick description would be helpful

Author: Added how to view outputs in App

## Supported Models

### Components
* `viam:luxonis:oak-d` - Configures the [OAK-D](https://shop.luxonis.com/products/oak-d) integrated camera model.
* `viam:luxonis:oak-ffc-3p` - Configures the [OAK-FFC-3P](https://shop.luxonis.com/products/oak-ffc-3p) non-integrated model with flexible board and camera sensor support.

Other OAK family cameras may work with the either component model, but are not officially supported and tested.

### Services
* `viam:luxonis:yolo-detection-network` - Configures [a DepthAI Yolo Detection Network](https://docs.luxonis.com/software/depthai-components/nodes/yolo_detection_network/) on the OAK device pipeline.

## How to use on your own machine
> [!NOTE]  
> Before configuring your resource, you must create a new smart machine. See [here](https://docs.viam.com/use-cases/configure/#configure-a-machine) for how to do that and more on the setup and configuration process as a whole.

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com/).

> [Add `camera` / `oak`,  `camera` / `oak-d` or `camera` / `oak-ffc-3p` to your machine](https://docs.viam.com/configure/#components).

Reviewer: ```suggestion
[Add `camera` / `oak-d` or `camera` / `oak-ffc-3p` to your machine](https://docs.viam.com/configure/#components).
```

Reviewer: since removing basic model!

Enter a name for your camera and click **Create**.

> To configure the vision service, click on the **Services** subtab and click **Create service**.

Reviewer: ```suggestion
To configure the vision service, add [``](/configure/#services).
```
This isn't the right instructions. What would be the name of the service? I can't figure out how to add it -- doesn't seem like it's a part of this module?

Author: Unfortunately the new version that actually does all these changes is not released yet.

Author: Try it again! I just released a new version

Author: Also the suggested change `To configure the vision service, add [``](/configure/#services).` results in this 
<img width="341" alt="Screenshot 2024-08-13 at 4 30 19 PM" src="https://github.com/user-attachments/assets/9a9abca0-6c6d-4ac2-a5e1-467efb2a9512">

# OAK Module

> This is a [Viam module](https://docs.viam.com/manage/configuration/#modules) for the [OAK](https://shop.luxonis.com/collections/oak-cameras-1) family of cameras. Registered at https://app.viam.com/module/viam/oak. The module supports getting color and stereo depth frames from OAK cameras using the `oak-d` and `oak-ffc-3p` component models, as well as setting up a YOLO detection network on the camera VPU using `yolo-detection-network` vision service model.

Author: The basic agnostic model is being deprecated

> This is a [Viam module](https://docs.viam.com/manage/configuration/#modules) for the [OAK](https://shop.luxonis.com/collections/oak-cameras-1) family of cameras. Registered at https://app.viam.com/module/viam/oak. The module supports getting color and stereo depth frames from OAK cameras using the `oak-d` and `oak-ffc-3p` component models, as well as setting up a YOLO detection network on the camera VPU using `yolo-detection-network` vision service model.

Reviewer: sounds good-- editing suggestion to reflect

* `viam:luxonis:oak-d` - Configures the [OAK-D](https://shop.luxonis.com/products/oak-d) integrated camera model.
* `viam:luxonis:oak-ffc-3p` - Configures the [OAK-FFC-3P](https://shop.luxonis.com/products/oak-ffc-3p) non-integrated model with flexible board and camera sensor support.

Other OAK family cameras may work with the either component model, but are not officially supported and tested.

### Services
* `viam:luxonis:yolo-detection-network` - Configures [a DepthAI Yolo Detection Network](https://docs.luxonis.com/software/depthai-components/nodes/yolo_detection_network/) on the OAK device pipeline.

## How to use on your own machine
> [!NOTE]  
> Before configuring your resource, you must create a new smart machine. See [here](https://docs.viam.com/use-cases/configure/#configure-a-machine) for how to do that and more on the setup and configuration process as a whole.

Navigate to the **Config** tab of your machine's page in [the Viam app](https://app.viam.com/).

For camera components, click on the **Components** subtab and click **Create component**.
Select the `camera` type, then select the `oak-d` or `oak-ffc-3p` model.
Enter a name for your camera and click **Create**.

> For vision services, click on the **Services** subtab and click **Create service**.

Reviewer: ```suggestion
To configure the vision service, click on the **Services** subtab and click **Create service**.
```

| `type` | str | Required | The type of the sensor: "color" or "depth", corresponding to whether the sensor on the respective socket is a color or mono depth camera. |
| `width_px` | int | Optional | Width in pixels of the images output by this camera. Default: `1280` |
| `height_px` | int | Optional | Height in pixels of the images output by this camera. Default: `720` |
| `frame_rate` | int | Optional | The frame rate the camera will capture images at. Default: `30` |
| `color_order` | str | Optional | The color order of the output frames (used for color sensor type only): "rgb" or "bgr". Default: `rgb` |
| `interleaved` | bool | Optional | Whether or not output frames should be stored in an interleaved format. Default: `false` |

> [!NOTE]  
> Higher resolutions may cause out of memory errors. See Luxonis documentation [here](https://docs.luxonis.com/projects/api/en/latest/tutorials/ram_usage/.).

> ### Configuring the YOLO Detection Network service model

Author: Done
